utils/log/progress.go

```diff
@@ -0,0 +1,292 @@
+}
+
+func initProxyReader(unit *progressBarUnit, reader io.Reader) io.ReadCloser {
+	if reader == nil {
+		return nil
+	}
+	rc, ok := reader.(io.ReadCloser)
+	if !ok {
+		rc = ioutil.NopCloser(reader)
+	}
+	return &proxyReader{unit, rc}
+}
+
+// Wraps an io.Reader for bytes reading
+type proxyReader struct {
+	unit *progressBarUnit
+	io.ReadCloser
+}
+
+// Overrides the Read method of the original io.Reader.
+func (pr *proxyReader) Read(p []byte) (n int, err error) {
+	n, err = pr.ReadCloser.Read(p)
+	if n > 0 {
+		pr.unit.incrChannel <- n
+	}
+	return
+}
+
+// Aborts a progress bar.
+// Should be called even if bar completed successfully.
+// The progress component's Abort method has no effect if bar has already completed, so can always be safely called anyway
+func (p *progressBarManager) Abort(barId int) {
+	p.barsRWMutex.RLock()
+	defer p.barsWg.Done()
+
+	// If a replacing bar
+	if p.bars[barId-1].replaceBar != nil {
+		// The replacing bar is displayed only if the replacedBar completed, so needs to be dropped only if so
+		if p.bars[barId-1].replaceBar.Completed() {
+			p.bars[barId-1].bar.Abort(true)
+		} else {
+			p.bars[barId-1].bar.Abort(false)
+		}
+	} else {
+		close(p.bars[barId-1].incrChannel)
+		p.bars[barId-1].bar.Abort(true)
+	}
+	p.barsRWMutex.RUnlock()
+}
+
+// Quits the progress bar while aborting the initial bars.
+func (p *progressBarManager) Quit() {
+	if p.headlineBar != nil {
+		p.headlineBar.Abort(true)
+		p.barsWg.Done()
+	}
+	if p.logFilePathBar != nil {
+		p.barsWg.Done()
+	}
+	// Wait a refresh rate to make sure all aborts have finished
+	time.Sleep(progressRefreshRate)
+	p.container.Wait()
+}
+
+// Initializes progress bar if possible (all conditions in 'shouldInitProgressBar' are met).
+// Creates a log file and sets the Logger to it. Caller responsible to close the file.
+// Returns nil, nil, err if failed.
+func InitProgressBarIfPossible() (log.ProgressBar, *os.File, error) {
+	shouldInit, err := shouldInitProgressBar()
+	if !shouldInit || err != nil {
+		return nil, nil, err
+	}
+
+	logFile, err := CreateLogFile()
+	if err != nil {
+		return nil, nil, err
+	}
+	log.SetLogger(log.NewLogger(GetCliLogLevel(), logFile))
+
+	newProgressBar := &progressBarManager{}
+	newProgressBar.barsWg = new(sync.WaitGroup)
+
+	// Initialize the progressBar container with wg, to create a single joint point
+	newProgressBar.container = mpb.New(
+		mpb.WithOutput(os.Stderr),
+		mpb.WithWaitGroup(newProgressBar.barsWg),
+		mpb.WithWidth(progressBarWidth),
+		mpb.WithRefreshRate(progressRefreshRate))
+
+	// Add headline bar to the whole progress
+	newProgressBar.printLogFilePathAsBar(logFile.Name())
+	newProgressBar.newHeadlineBar(" Working... ")
+
+	return newProgressBar, logFile, nil
+}
+
+// Init progress bar if all required conditions are met:
+// CI == false (or unset), Stderr is a terminal, and terminal width is large enough
+func shouldInitProgressBar() (bool, error) {
+	isCI, err := utils.GetBoolEnvValue("CI", false)
```

Comment:
It is better to use **ci** instead of **isCI** according to the common code conversions.